library/alloc/src/vec/mod.rs

```diff
@@ -1231,6 +1237,16 @@ impl<T, A: Allocator> Vec<T, A> {
     /// }
     /// ```
     ///
+    /// The validity guarantee works out this way:
+    ///
+    /// ```rust
+    /// let mut v = vec![0];
+    /// let ptr = v.as_ptr();
+    /// let x = ptr.read();
+    /// v[0] = 5;
```

Comment:
i tried to write a similar example for `as_ptr` and wanted to avoid using `as_mut_ptr()`, but I think this is _also_ unsound, yes? @RalfJung

Reply:
Whether or not you can write through a pointer from `Vec::as_ptr` is up to the standard library implementation. Miri does not complain about this program:
```rust
fn main() {
    let v = vec![0];
    unsafe {
        *(v.as_ptr() as *mut _) = 1;
    }
}
```

Reply:
Oh that's not at all related to what you were asking. Oops.

Reply:
I would say it is up to t-libs if they want to guarantee this -- basically this means that the pointer returned here was not derived via a shared reference. By analogy with `addr_of!`, I think we want to allow this code. (Is it worth mentioning that analogy?)